classes/audit_manager.php

```diff
@@ -0,0 +1,376 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * @package     report_apocalypse
+ * @author      Tom Dickman <tomdickman@catalyst-au.net>
+ * @copyright   2019 Catalyst IT Australia {@link http://www.catalyst-au.net}
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace report_apocalypse;
+
+defined('MOODLE_INTERNAL') || die;
+
+use stdClass;
+use moodle_url;
+use DateTime;
+
+/**
+ * Manager class for running audits and instantiating audit_activities.
+ *
+ * @package report_apocalypse
+ */
+class audit_manager implements audit_interface {
+
+    /**
+     * Run a flash audit on the moodle site.
+     *
+     * @throws \dml_exception
+     * @throws \dml_transaction_exception
+     */
+    public static function run_audit() {
+        global $DB;
+
+        $modules = self::get_all_modules();
+        list($sql, $params) = self::build_sql_and_parameters_for_audit($modules);
+        $recordset = $DB->get_recordset_sql($sql, $params);
+
+        $transaction = $DB->start_delegated_transaction();
+        try {
+            self::delete_records();
+            self::store_records($recordset);
+
+            $transaction->allow_commit();
+        } catch (Exception $e) {
+            $transaction->rollback($e);
+        }
+    }
+
+    /**
+     * Get all audit results as a moodle_recordset.
+     *
+     * @return \moodle_recordset A moodle_recordset instance
+     * @throws \dml_exception
+     */
+    public static function get_audit_results() {
```

Comment:
this shouldn't be a function - you also need to make sure that when you use recordsets you must close them in the same function - you can't open a recordset in one function and then close it in another... but this really shouldn't be a function on it's own in the first place.

Reply:
Removed.